File: tests/test_e2e.js
const fs = require('fs');
const path = require('path');

const API_URL = 'http://localhost:8000';

async function test_pipeline(filepath) {
    console.log(`\n--- Testing ${filepath} ---`);
    if (!fs.existsSync(filepath)) {
        console.log(`File not found: ${filepath}`);
        return;
    }
    
    try {
        const fileData = fs.readFileSync(filepath);
        const formData = new FormData();
        const blob = new Blob([fileData], { type: 'application/vnd.tcpdump.pcap' });
        formData.append('file', blob, path.basename(filepath));

        // 1. Upload
        const uploadRes = await fetch(`${API_URL}/upload_pcap`, { method: 'POST', body: formData });
        if (!uploadRes.ok) {
            console.log(`Upload failed: ${uploadRes.status} ${await uploadRes.text()}`);
            return;
        }
        const ipsecReq = await uploadRes.json();

        // 2. Assess
        const assessRes = await fetch(`${API_URL}/assess`, {
            method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(ipsecReq)
        });
        const risk = await assessRes.json();

        // 3. Remediate
        const remediateRes = await fetch(`${API_URL}/remediate`, {
            method: 'POST', headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                risk_label: risk.risk_label,
                flagged_issues: risk.flagged_issues,
                top_contributing_factors: risk.top_contributing_factors
            })
        });
        const remediation = await remediateRes.json();

        // 4. PQC
        const pqcRes = await fetch(`${API_URL}/pqc_score`, {
            method: 'POST', headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                encryption_algorithm: ipsecReq.encryption_algorithm,
                key_length_bits: ipsecReq.key_length_bits,
                hash_algorithm: ipsecReq.hash_algorithm,
                dh_group: ipsecReq.dh_group
            })
        });
        const pqc = await pqcRes.json();

        // 5. Log
        const logRes = await fetch(`${API_URL}/log`, {
            method: 'POST', headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                report_data: { ipsec_request: ipsecReq, risk_assessment: risk, pqc_score: pqc.pqc_score }
            })
        });
        const audit = await logRes.json();

        console.log("Success!");
        console.log("Risk Score:", risk.risk_score, risk.risk_label);
        console.log("PQC Safe:", pqc.is_quantum_safe);
        console.log("Remediation length:", remediation.config_diff?.length);
        console.log("Audit Hash:", audit.report_hash);
    } catch (e) {
        console.log(`Exception: ${e}`);
    }
}

async function runTests() {
    await test_pipeline("scenario_critical_legacy.pcap");
    await test_pipeline("scenario_moderate_transition.pcap");
    await test_pipeline("scenario_strong_modern.pcap");

    fs.writeFileSync("dummy.pcap", "NOT A REAL PCAP DATA".repeat(10));
    await test_pipeline("dummy.pcap");

    console.log("\n--- Testing Audit Verification ---");
    try {
        const verify_resp = await fetch(`${API_URL}/verify`);
        console.log(`Verification Status: ${verify_resp.status}`);
        console.log(await verify_resp.json());
    } catch (e) {
        console.log(`Exception: ${e}`);
    }
}

runTests();

File: tests/test_e2e.py
import time
from fastapi.testclient import TestClient
from backend.main import app
import json

client = TestClient(app)

def run_pipeline(pcap_filename, expected_risk_label):
    print(f"\n--- Testing Scenario: {pcap_filename} ---")
    start_time = time.time()
    
    # 1. Upload PCAP
    with open(pcap_filename, "rb") as f:
        res_upload = client.post("/upload_pcap", files={"file": (pcap_filename, f, "application/vnd.tcpdump.pcap")})
    assert res_upload.status_code == 200, f"Upload failed: {res_upload.text}"
    ipsec_req = res_upload.json()
    print("[OK] Upload & Parse Successful")
    
    # 2. Assess
    res_assess = client.post("/assess", json=ipsec_req)
    assert res_assess.status_code == 200, f"Assess failed: {res_assess.text}"
    risk = res_assess.json()
    assert risk["risk_label"] == expected_risk_label, f"Expected {expected_risk_label}, got {risk['risk_label']}"
    print(f"[OK] Assess Successful (Label: {risk['risk_label']}, Score: {risk['risk_score']})")
    
    # 3. PQC Score
    res_pqc = client.post("/pqc_score", json={
        "encryption_algorithm": ipsec_req["encryption_algorithm"],
        "key_length_bits": ipsec_req["key_length_bits"],
        "hash_algorithm": ipsec_req["hash_algorithm"],
        "dh_group": ipsec_req["dh_group"]
    })
    assert res_pqc.status_code == 200, "PQC failed"
    pqc = res_pqc.json()
    assert "pqc_score" in pqc, "pqc_score missing from response"
    print(f"[OK] PQC Successful (Score: {pqc['pqc_score']})")
    
    # 4. Remediate
    res_rem = client.post("/remediate", json={
        "risk_label": risk["risk_label"],
        "flagged_issues": risk["flagged_issues"],
        "top_contributing_factors": risk["top_contributing_factors"]
    })
    assert res_rem.status_code == 200, "Remediate failed"
    rem_data = res_rem.json()
    config_diff = rem_data.get("config_diff", "")
    assert "! No configuration available" not in config_diff, "LLM returned the fallback string silently!"
    assert len(config_diff) > 20, "Remediation config is suspiciously short."
    assert any(k in config_diff.lower() for k in ["crypto", "isakmp", "ipsec", "transform-set", "ikev2", "proposal", "policy"]), "Remediation config lacks valid IPsec keywords."
    print("[OK] LLM Remediation Successful (Dynamic content verified)")
    print(f"\n--- GENERATED CONFIG DIFF ---\n{config_diff}\n-----------------------------\n")
    
    # 5. Log
    res_log = client.post("/log", json={"report_data": ipsec_req})
    assert res_log.status_code == 200, "Log failed"
    audit = res_log.json()
    assert "report_hash" in audit, "report_hash missing from response"
    print(f"[OK] Audit Trail Successful (Hash: {audit['report_hash']})")
    
    end_time = time.time()
    elapsed = end_time - start_time
    print(f"Pipeline completed in {elapsed:.2f} seconds.")
    assert elapsed < 45.0, "Performance threshold exceeded (>45s)"
    return True

def test_dummy_pcap():
    print("\n--- Testing Scenario: dummy.pcap ---")
    # Create a dummy pcap file
    with open("samples/dummy.pcap", "wb") as f:
        f.write(b"this is not a valid pcap file")
    
    with open("samples/dummy.pcap", "rb") as f:
        res = client.post("/upload_pcap", files={"file": ("samples/dummy.pcap", f, "application/vnd.tcpdump.pcap")})
    
    assert res.status_code == 400, "Dummy PCAP should be rejected"
    print(f"[OK] Dummy PCAP gracefully rejected with: {res.json()['detail']}")
    import os
    os.remove("samples/dummy.pcap")

if __name__ == "__main__":
    run_pipeline("samples/scenario_critical_legacy.pcap", "Critical")
    run_pipeline("samples/scenario_moderate_transition.pcap", "Weak")
    run_pipeline("samples/scenario_strong_modern.pcap", "Strong")
    test_dummy_pcap()
    print("\nALL SCENARIOS PASSED E2E SMOKE TEST.")



